### ISSUE TITLE
Abnormal loss value during training the m4c captioner on COCO and Textcaps for the Image Captioning. 

### ISSUE THREAD

**Issue by TownWilliam** [facebookresearch/mmf#949]
## ❓ Questions and Help
When I run the m4c captioner model on the dataset Textcaps for the Imgae captioning task, I find the loss value output in report is greatly numerous-millions or so. The total loss is strange but the final trainning result on metrics liking BLEU-$ is basically right. A unexpected loss gets an expected results. So my question is that: what's wrong with it? The segment of the log is next:

[32m2021-03-31T01:11:08 | mmf.trainers.callbacks.logistics: [0mprogress: 31900/32000, **train/textcaps/m4c_decoding_bce_with_mask: 264317184.0000**, train/textcaps/m4c_decoding_bce_with_mask/avg: 304672683.2169, train/total_loss: 264317184.0000, train/total_loss/avg: 304672683.2169, max mem: 7019.0, experiment: run, epoch: 6, num_updates: 31900, iterations: 31900, max_updates: 32000, lr: 0., ups: 0.19, time: 08m 39s 014ms, time_since_start: 36h 13m 18s 560ms, eta: 08m 45s 243ms
[32m2021-03-31T01:20:26 | mmf.trainers.callbacks.checkpoint: [0mCheckpoint time. Saving a checkpoint.

Later, I debug the model and find that it use BCEwithMaskLoss and the error is caused by a 1e10 or so loss value in the result.    
The condition is as follows:
![图片1](https://user-images.githubusercontent.com/49356039/118813544-d5db3f00-b8e1-11eb-8484-d124a4dceed8.png)

Then, I find the loss value is produced by a greatly large predict scores: -1e10. Even it then was normolized to [0,1], it is still very tiny. So the loss value is still very big.
![image](https://user-images.githubusercontent.com/49356039/118814378-aed13d00-b8e2-11eb-931c-81e05b24eb2e.png)

It seems that this strange number value is numerical overflow. The abnormal score is directly calculated by the y=Wx_{dec}+b. Personally I think it might be  something wrong with the scores prediction network weight.
Now I want to konw how could I solve the problem? Or it could be called as a "problem"?

I am looking forward  for your reply. Thanks a lot.

**Comment by ronghanghu** [comment]
Hi @TownWilliam, sorry for the confusion here! This happens because the score of <unk> token (the token for unknown word) was hard-coded to -1e10 in the model, such that the model will never produce <unk> as an answer at test time. So when the ground-truth answer is an unknown word, it will produce a very large loss. This behavior is introduced here: https://github.com/facebookresearch/mmf/blob/master/mmf/models/m4c_captioner.py#L19-L21. One can also turn off this behavior and allow the model to train on <unk> 

The training and testing should be fine in the current codebase. One can also turn off this behavior by changing https://github.com/facebookresearch/mmf/blob/master/mmf/configs/models/m4c_captioner/defaults.yaml#L28 to "remove_unk_in_pred: false", which should give similar results.

**Comment by TownWilliam** [comment]
@ronghanghu Thanks a lot!   This details is ignored in M4C Captioner. 

By the way, I wonder why does the process of back propagation is normal despite the abnormal when calculating the gradient.
The so huge loss value should influence the gradient in the back propagation. However, it does not.

**Comment by ronghanghu** [comment]
Hi @TownWilliam, the score of `<unk>` token (the token for unknown word) was hard-coded to a dummy value of -1e10. This gives a large gradient only to the `<unk>` score using BinaryCrossEntropy loss. But since the `<unk>` score is a dummy score, it won't get back propagated to the neural network parameters.

(As mentioned above, one can also turn off this behavior by changing https://github.com/facebookresearch/mmf/blob/master/mmf/configs/models/m4c_captioner/defaults.yaml#L28 to "remove_unk_in_pred: false", which should give similar results.)

**Comment by TownWilliam** [comment]
@ronghanghu Thanks a lot. You solve my confusion.
